test/common/upstream/upstream_impl_test.cc

```diff
@@ -1785,6 +1806,57 @@ TEST_F(ClusterInfoImplTest, ExtensionProtocolOptionsForFilterWithOptions) {
   }
 }
 
```

Comment:
I'm honestly not happy about these unit tests. ClusterInfoImpl is fairly high level, yet I'm depending on the low level syscall interactions to prove that I have the correct socket options.

I considered moving the actual syscall into the Socket class itself, but I wasn't sure of the original design intent of the SocketOptions listener -- i.e. whether it was to abstract the details of that syscall, or whether it was to just abstract the socket options themselves.

Another alternative would be to add some introspection to the socket options class, but again somewhat defeats their purpose, since the socket options are platform specific, vary in size, etc. If anyone has suggestions for how to improve this without too much pain, I'm all ears. :)

Reply:
Hm. I don't think there's a better way right now. Perhaps it makes sense for Network::Socket::Option to provide a toString() method that implementations could use to describe themselves? Then you could just iterate the cluster options and make sure the expected string was provided and leave testing the actual system call to the socket option tests. The existing implementations could just emit convert the level, option and value in some formatted string.

I think the alternative is either keeping this deep mocking of syscalls or else providing a SocketOptionName interface that would only really be used in tests. But, I think it would end up generating a similar string from the option data.

@htuch does that seem reasonable?

Reply:
We should probably have some feature in `Socket::Option` to map back to the abstract socket description; it doesn't need to be a string though, it could be `SocketOptionName` etc.

Reply:
I think SocketOptionName is a std::pair<int, int> right now. That seems like it might not match up well with VPP and/or QUIC? Or does that not matter?

Reply:
If we're using a layer around VPP to provide a socket API abstraction, I'd hope that it also provides the same api for options. I'm not sure about quick, though--do typical implementations pass through socket options to the underlying UDP socket?

Reply:
I expect they'll all support options (perhaps not all the same ones), but it's not clear to me if they'll be describable as pairs of ints? I suppose we can just move SocketOptionName to the API for now and then rejigger it when it's understood what option names for the other kinds of sockets look like.

Reply:
That's probably fine, and it'll solve at least one of the complexities: show that the option is set at all. However, one concern I have is that we'll lose coverage of the actual value being set for non "flag" options, like SO_MARK. The name tells us that the option is present, but not its actual value. Are we really okay with that?

Reply:
@zuercher @htuch 

Any thoughts on my last comment? Is this sufficient, given that we won't get perfect coverage? Leave that to component tests, maybe?

Reply:
I think I misunderstood somewhere along the way. I thought we were just going to provide a way to get the SocketOptionName from Socket::Option. That doesn't describe the value; it only has a pair of ints for the family and the option. 

The value itself is an `int`. I think we'd probably be fine returning a pair of <SocketOptionName, int>, or even <SoketOptionName, std::string> to describe it. If it's mostly for introspection, the cost of the string conversion shouldn't be prohibitive, and it *is* flexible.

So, what about adding something like the following to Socket::Option:

```
public:
  virtual std::pair<SocketOptionaName, std::string> getInformation() const PURE;
```

I'm open to suggestions on the naming. :)

Reply:
My concern was primarily that the shape of SocketOptionName might not work for all socket implementations, including some that we expect to be added soon. I'm ok just moving SocketOptionName to the API as-is and then modifying it later if turns out to be inadequate. 

A somewhat hackier way might be to just modify SocketOptionImpl to provide access to the name/value and then just iterate over the OptionsSharedPtr values, casting them to SocketOptionImpl to verify that the correct option/value are set.

Reply:
I'm good with either of these.

Reply:
A bit of a hitch. Some implementations have more than one set of options to apply -- e.g. the AddrFamilyAwareSocketOptionImpl. That means any interface which assumes a single value, like "getInformation()" will fail.

So, we need to pass in the "visitor" options to fully mimic the production interface. I guess I'll just do that.

I don't really like doing the cast if I can avoid it... I'm not sure it's any better than the test I have already. :)

Reply:
Went with a function to return an optional information struct. Arguments are the socket and the state. If the option wouldn't be applied in that case, it returns nullopt. Otherwise, it packages up the info, and returns that.